pytester: use `monkeypatch.chdir()` for dir changing

bluetech opened the pull request:
The current method as the following problem, described by @sadra-barikbin in https://github.com/pytest-dev/pytest/pull/11236#issuecomment-1666851689:

The tests that request both `pytester` and `monkeypatch` and use `monkeypatch.chdir` without context, relying on `monkeypatch`'s teardown to restore cwd. This doesn't work because the following sequence of actions take place:

- `monkeypatch` is set up.
- `pytester` is set up. It saves the original cwd and changes it to a new one dedicated to the test function.
- Test function calls `monkeypatch.chdir()` without context. `monkeypatch` saves cwd, which is not the original one, before changing it.
- `pytester` is torn down. It restores the cwd to the original one.
- `monkeypatch` is torn down. It restores cwd to what it has saved.

The solution here is to have pytester use `monkeypatch.chdir()` itself, then everything is handled correctly.

<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [ ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->

RonnyPfannschmidt commented on src/_pytest/pytester.py:
```diff
@@ -696,15 +688,14 @@ def __init__(
         #: the method using them so refer to them for details.
         self.plugins: List[Union[str, _PluggyPlugin]] = []
-        self._cwd_snapshot = CwdSnapshot()
         self._sys_path_snapshot = SysPathsSnapshot()
```
should we log a followup to ensure sys.path/sys.modules handling in coordination with monkeypatch

bluetech replied:
Might be possible but some complications:

- `SysPathSnapshot` also saves `sys.meta_path`, monkeypatch doesn't.
- Monkeypatch doesn't save `sys.path` by default, though pytester can probably force it to do it
- `SysModulesSnapshot` has no equivalent in monkeypatch currenty as far as know. Also has some `preserve_modules` exclusion...

bluetech commented:
Updated; there were some calls to `monkeypatch.undo()` in the test suite which interfered with this change (and theoretically, with other things even before this change). I changed them to use `monkeypatch.context()`.

nicoddemus commented on testing/code/test_excinfo.py:
```diff
@@ -866,14 +870,14 @@ def entry():
-        last_reprtb = last_p.repr_traceback_entry(excinfo.traceback[-1], excinfo)
-        last_lines = last_reprtb.lines
-        monkeypatch.undo()
+        with monkeypatch.context() as mp:
```
Because others use `pytester`, I think it would be worth adding a `trivial` changelog explaining this change, in case others find/have the same pattern in their test suites.

bluetech replied:
Agreed - added a changelog.

RonnyPfannschmidt commented:
Should we deprecate monkeypatch.undo as public api now that . context is available

bluetech commented:
We did add [a note](https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.undo) warning about its use some time ago. I don't know if we can entirely deprecate it because it doesn't have a replacement if someone wants to undo their own `MonkeyPatch` instance.

RonnyPfannschmidt commented:
@bluetech customized contextmanager undo is controlled via `ExitStack`
